feature/staimov_kafka_3_with_checksum

staimov opened the pull request:
Домашнее задание 3. 
В этой версии в качестве подтверждения обработки транзакций используется контрольная сумма за интервал времени

staimov committed:
kafka_3: подтверждения теперь содержат контрольную сумму за интервал времени + другие мелкие исправления

staimov committed:
kafka_3: minor fix

staimov committed:
kafka_3: minor fix

staimov committed:
kafka_3: minor fix

staimov committed:
kafka_3: refactor packages

staimov committed:
Merge remote-tracking branch 'origin/feature/staimov_kafka_3' into feature/staimov_kafka_3

Liman26 commented on src/main/java/sbp/school/kafka/ack/service/AckConsumer.java:
```diff
@@ -91,20 +93,36 @@ private void onCommitComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Ex
+                transactionProducer.cleanupInterval(intervalKey);
+                logger.debug("Получено и обработано подтверждение: intervalKey={}, offset={}", intervalKey, record.offset());
+            } else {
+                logger.warn("Получено подтверждение, однако контрольная сумма не совпадает: " +
```
Просто логируем? Переотправку инициировать не будем?

staimov replied:
Здесь в методе sbp.school.kafka.ack.service.AckConsumer#processRecord происходит только фиксация и логирование подтверждений. 
Переотправка транзакций, для которых истек таймаут получения валидного подтверждения, происходит в отдельном потоке и реализована в методе sbp.school.kafka.transaction.service.TransactionProducer#retrySendTransactions

Liman26 commented on src/main/java/sbp/school/kafka/transaction/service/TransactionProducer.java:
```diff
@@ -0,0 +1,170 @@
+                .collect(Collectors.toSet());
+
+        for (Long intervalKey : internalKeysToRetry) {
+            // TODO: возможно надо еще ограничивать число повторных отправок
```
Обязательно

staimov replied:
Добавил счетчик и ограничение количества переотправок (https://github.com/staimov/JavaSchool/pull/5/commits/8842812b8a48bbfda4ad189add2aeb020ca65d9d)

Liman26 commented on src/main/java/sbp/school/kafka/ack/service/AckConsumer.java:
```diff
@@ -91,20 +93,36 @@ private void onCommitComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Ex
-        }
-        if (!transactionProducer.getUnackedTransactions().containsKey(ackId)) {
-            logger.warn("Пропущено неизвестное подтверждение: id={}, offset={}", ackId, record.offset());
+    private void processRecord(ConsumerRecord<String, AckDto> record) {
```
Что будет, если сообщений с подтверждениями вообще не будет? Мы узнаем об этом?

staimov replied:
Да, если мы вообще не получим подтверждения, то для пачки транзакций, для которой истечет таймаут получения подтверждения, будет выполнена переотправка.
Проверка необходимости переотправки происходит в отдельном потоке и реализована в методе sbp.school.kafka.transaction.service.TransactionProducer#retrySendTransactions

staimov commented on src/main/java/sbp/school/kafka/Main.java:
```diff
@@ -3,6 +3,6 @@
     public static void main(String[] args) {
         // Чтобы проверить работу запусти тест
-        // sbp.school.kafka.producer.TransactionProducerTest.testSendMultipleTransactions_And_WaitForAcks_And_RetrySend
+        // sbp.school.kafka.transaction.TransactionProducerTest.testSendMultipleTransactions_And_WaitForAcks
```
Это не закомментированный код, а просто комментарий, поясняющий, почему метод main в данном случае пустjq

staimov committed:
kafka_3:
добавлен счетчик повторных отправок;
добавлено ограничение на количество повторных отправок;
добавлен сервис хранения информации об отправленных транзакциях

staimov commented on src/main/java/sbp/school/kafka/Main.java:
```diff
@@ -3,6 +3,6 @@
     public static void main(String[] args) {
         // Чтобы проверить работу запусти тест
-        // sbp.school.kafka.producer.TransactionProducerTest.testSendMultipleTransactions_And_WaitForAcks_And_RetrySend
+        // sbp.school.kafka.transaction.TransactionProducerTest.testSendMultipleTransactions_And_WaitForAcks
```
Это не закомментированный код, а просто комментарий, поясняющий, почему метод sbp.school.kafka.Main#main в данном случае пустой

staimov committed:
kafka_4: fix

staimov committed:
kafka_4: tests updated
